drivers/wdm/audio/backpln/portcls/miniport_dmus.cpp

```diff
@@ -67,12 +67,11 @@ const ULONG      kMPUInputBufferSize = 128;
  * reference counting and aggregation support.
  */
 class CMiniportDMusUART
-:   public IMiniportDMus,
-    public IMusicTechnology,
-    public IPowerNotify
+:   public CUnknownImpl<IMiniportDMus,
+                        IMusicTechnology,
+                        IPowerNotify>
```

Comment:
So that I understand the need of the (C++11 😡) variadic template, is it because if there is inheritance from other classes implementing these interfaces, they will share the same reference counter?

Reply:
Yep. If you inherited from three different `CUnknownImpl`s, you'd get three refcounts. If you only make the first interface a `CUnknownImpl`, the other interfaces would be lacking AddRef/Release implementations.
Happy to solve this another way if there is one -- but the only one I found was to put the implementation behind a macro. Or I guess you could have `CUnknownImpl1`, `CUnknownImpl2`, `CUnknownImpl3`, but that's pretty ridiculous, too.

Reply:
Would using virtual inheritance be another way?
`class CFooBar : virtual public IInterface1, virtual public IInterface2`

Reply:
As far as I can tell, that doesn't help -- https://godbolt.org/z/soMfn396c
But if you can get it to work in that snippet somehow, we can probably get it to work in the real world, too.

Reply:
I got it working by adding some missing "virtual" to the inheritances of the intermediate classes: see https://godbolt.org/z/dorxxcz1n

Reply:
I would love if we could keep the compatibility to the other compilers here!